### PR TITLE
Introduce alter view propagation

### PR DESCRIPTION
**Pull request by velioglu**
DESCRIPTION: Adds support for propagating ALTER VIEW commands

Introduce alter view propagation

Adds support for propagating alter view commands (except setting/resetting default values as Citus doesn't support update/delete on views). Note that both ALTER VIEW and ALTER TABLE supports have been added, since PG supports both of those syntax to alter view.

### CODE REVIEW

File: src/backend/distributed/commands/table.c [hunk 1]
```diff
@@ -652,11 +652,17 @@ PostprocessAlterTableSchemaStmt(Node *node, const char *queryString)
...
 
 	/* check whether we are dealing with a sequence here */
-	if (get_rel_relkind(tableAddress.objectId) == RELKIND_SEQUENCE)
+	char relKind = get_rel_relkind(tableAddress.objectId);
+	if (relKind == RELKIND_SEQUENCE)
 	{
 		stmt->objectType = OBJECT_SEQUENCE;
 		return PostprocessAlterSequenceSchemaStmt((Node *) stmt, queryString);
 	}
+	else if (relKind == RELKIND_VIEW)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:661]: Q1: why is this happening before `!ShouldPropagate` ?  

Q2: Also, we have `PostprocessAlterViewSchemaStmt` as well, why do we have this code? Is it because we can do `ALTER TABLE view SET SCHEMA`? Even that seems to go through `PostprocessAlterViewSchemaStmt`

I'm in favor of not allowing such weird syntax even if PG allows. As one of the most senior PG commiters noted for PG [here](https://www.postgresql.org/message-id/flat/F681D64D-8C81-4665-8ABE-D8DCF1A8E511%40amazon.com#57e747c1f3b7904d44fb0a5ee7a296e8):

> In short: no, I do not agree that this is a bug to be fixed.  Perhaps
we should have done things differently years ago, but it's too late to
redefine it.

I think it is time for us to build things in a way we prefer. We shouldn't really need to support this weird approach. Isn't it confusing?

Comment (onderkalaci) [reply src/backend/distributed/commands/table.c:661]: > Q1: why is this happening before !ShouldPropagate ?

because `PostprocessAlterViewSchemaStmt` does that anyway.

File: src/backend/distributed/deparser/deparse_view_stmts.c [hunk 2]
```diff
@@ -92,3 +99,206 @@ AppendViewNameList(StringInfo buf, List *viewNamesList)
...
+		}
+
+		case AT_ResetRelOptions:
+		{
+			AppendAlterViewResetOptionsStmt(buf, alterTableCmd);
+			break;
+		}
+
+		case AT_ColumnDefault:
+		{
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_view_stmts.c:159]: do we throw errors in these deparser code or earlier in general? This seems a little too late to throw the error?

File: src/backend/distributed/commands/distribute_object_ops.c [hunk 3]
```diff
@@ -1512,6 +1546,23 @@ GetDistributeObjectOps(Node *node)
 					return &Trigger_Rename;
 				}
 
+				case OBJECT_VIEW:
+				{
+					return &View_Rename;
+				}
+
+				case OBJECT_COLUMN:
```
Comment (onderkalaci) [inline src/backend/distributed/commands/distribute_object_ops.c:1554]: This is a little confusing for me to follow. why we need this block, does pg support rename column of a view? Maybe, can you remind me which command triggers this code-path?

File: src/test/regress/expected/view_propagation.out [hunk 4]
```diff
@@ -359,13 +359,93 @@ ERROR:  cannot change name of view column "id" to "a"
...
+
+\c - - - :master_port
+SET search_path to view_prop_schema;
+ALTER TABLE alter_view_1 RESET (check_option, security_barrier);
+ALTER VIEW alter_view_1 RESET (check_option, security_barrier);
+-- Change the schema of the view
+ALTER TABLE alter_view_1 SET SCHEMA view_prop_schema_inner;
+ALTER VIEW view_prop_schema_inner.alter_view_1 SET SCHEMA view_prop_schema;
+-- Rename view and view's column name
+ALTER VIEW alter_view_1 RENAME COLUMN val1 TO val2;
```
Comment (onderkalaci) [inline src/test/regress/expected/view_propagation.out:423]: would be good to show that after `rename column` we can still de-parse the alias of the view properly with the new name. For example, we can re-sync the metadata (or only get the metadata events) and show  that alias is correct

**Review by onderkalaci — COMMENTED** [review]
Looks good overall, I need to test this a bit more. For now, few comments:
1) Can we avoid overriding the object types in the statements, but let the code handle both views and tables where necessary?
2) Can we send the exact commands that are executed on the coordinator (e.g., do not change `ALTER TABLE` to `ALTER VIEW ` on the deparser
3) Lets add the comments that we discussed earlier where we do the process redirection

File: src/backend/distributed/commands/distribute_object_ops.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/distribute_object_ops.c:1554]: e.g., why do we need the `else` block, it didn't exists before this PR?

File: src/backend/distributed/commands/table.c (hunk 1 above)
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:661]: as we noted, lets extend the comments where we `re-route` the commands

Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:661]: as we noted, lets extend the comments where we `re-route` the commands

File: src/backend/distributed/commands/table.c [hunk 5]
```diff
@@ -1758,18 +1772,29 @@ PreprocessAlterTableSchemaStmt(Node *node, const char *queryString,
 	{
 		return NIL;
 	}
+
+	QualifyTreeNode((Node *) stmt);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:1776]: as we discussed, would be nice to (a) make this change as a separate commit in case we might want to backport this to earlier versions (b) have a simple test for sequences as well

Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:1776]: as we discussed, would be nice to (a) make this change as a separate commit in case we might want to backport this to earlier versions (b) have a simple test for sequences as well

File: src/backend/distributed/commands/table.c [hunk 6]
```diff
@@ -1939,12 +1964,19 @@ PostprocessAlterTableStmt(AlterTableStmt *alterTableStatement)
...
+		char relKind = get_rel_relkind(relationId);
+		if (relKind == RELKIND_SEQUENCE)
 		{
 			AlterTableStmtObjType_compat(alterTableStatement) = OBJECT_SEQUENCE;
 			PostprocessAlterSequenceOwnerStmt((Node *) alterTableStatement, NULL);
 			return;
 		}
+		else if (relKind == RELKIND_VIEW)
+		{
+			AlterTableStmtObjType_compat(alterTableStatement) = OBJECT_VIEW;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:1976]: no action needed for now; it is an odd syntax, it would probably be nicer if we had `SetAlterTableStmtObjType_compat(alterTableStatement, OBJECT_VIEW)` or such

File: src/backend/distributed/commands/view.c [hunk 7]
```diff
@@ -416,3 +417,255 @@ CreateAlterViewOwnerCommand(Oid viewOid)
...
+	AlterTableStmt *stmt = castNode(AlterTableStmt, node);
+	Assert(AlterTableStmtObjType_compat(stmt) == OBJECT_VIEW);
+
+	ObjectAddress viewAddress = GetObjectAddressFromParseTree((Node *) stmt, false);
+	if (!ShouldPropagateObject(&viewAddress))
+	{
+		return NIL;
+	}
+
+	if (IsObjectAddressOwnedByExtension(&viewAddress, NULL))
```
Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:469]: as I noted in the other PR as well, lets have one view test owned by any extension. These are easy to break and hard to recognize without tests ;)  Same here for https://github.com/citusdata/citus/pull/5923#pullrequestreview-962050471

Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:469]: as I noted in the other PR as well, lets have one view test owned by any extension.

File: src/backend/distributed/commands/view.c [hunk 8]
```diff
@@ -416,3 +417,255 @@ CreateAlterViewOwnerCommand(Oid viewOid)
...
+{
+	AlterObjectSchemaStmt *stmt = castNode(AlterObjectSchemaStmt, node);
+
+	ObjectAddress typeAddress = GetObjectAddressFromParseTree((Node *) stmt, false);
+	if (!ShouldPropagateObject(&typeAddress))
+	{
+		return NIL;
+	}
+
+	QualifyTreeNode((Node *) stmt);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:596]: do we need ensure coordinator here? In general, why do we call `QualifyTreeNode` here as we are not using it here? e.g., can't we postpone this until we need in post process?

Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:596]: do we need ensure coordinator here?

File: src/backend/distributed/commands/view.c [hunk 9]
```diff
@@ -416,3 +417,255 @@ CreateAlterViewOwnerCommand(Oid viewOid)
...
+ */
+ObjectAddress
+AlterViewSchemaStmtObjectAddress(Node *node, bool missing_ok)
+{
+	AlterObjectSchemaStmt *stmt = castNode(AlterObjectSchemaStmt, node);
+
+	Oid viewOid = RangeVarGetRelid(stmt->relation, NoLock, true);
+
+	/*
+	 * Since it can be called both before and after local execution, we need to check
```
Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:647]: minor: `local execution` is overloaded term for Citus, maybe use `standardProcess utility` or such

Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:647]: minor: `local execution` is overloaded term for Citus, maybe use `standardProcess utility` or such

File: src/backend/distributed/commands/view.c [hunk 10]
```diff
@@ -416,3 +417,255 @@ CreateAlterViewOwnerCommand(Oid viewOid)
...
+/*
+ * AlterViewSchemaStmtObjectAddress returns the ObjectAddress of the view that is the object
+ * of the alter schema statement.
+ */
+ObjectAddress
+AlterViewSchemaStmtObjectAddress(Node *node, bool missing_ok)
+{
+	AlterObjectSchemaStmt *stmt = castNode(AlterObjectSchemaStmt, node);
+
+	Oid viewOid = RangeVarGetRelid(stmt->relation, NoLock, true);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:644]: should we pass `missing_ok` instead of `true`?

Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:644]: should we pass `missing_ok` instead of `true`?

File: src/backend/distributed/deparser/deparse_view_stmts.c [hunk 11]
```diff
@@ -92,3 +99,206 @@ AppendViewNameList(StringInfo buf, List *viewNamesList)
...
+
+		appendStringInfo(buf, "%s", def->defname);
+		if (def->arg != NULL)
+		{
+			appendStringInfo(buf, "=");
+			appendStringInfo(buf, "%s", defGetString(def));
+		}
+	}
+
+	appendStringInfo(buf, ")");
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_view_stmts.c:208]: do we need `;` at the end? Just in case we send multiple commands in one round-trip for some reason in the future. Maybe go over the commands and add if needed?

Comment (onderkalaci) [inline src/backend/distributed/deparser/deparse_view_stmts.c:208]: do we need `;` at the end? Just in case we send multiple commands in one round-trip for some reason in the future. Maybe go over the commands and add if needed?

File: src/backend/distributed/deparser/qualify_view_stmt.c [hunk 12]
```diff
@@ -50,3 +50,57 @@ QualifyDropViewStmt(Node *node)
...
+void
+QualifyRenameViewStmt(Node *node)
+{
+	RenameStmt *stmt = castNode(RenameStmt, node);
+	RangeVar *view = stmt->relation;
+
+	if (view->schemaname == NULL)
+	{
+		Oid schemaOid = RangeVarGetCreationNamespace(view);
+		view->schemaname = get_namespace_name(schemaOid);
```
Comment (onderkalaci) [inline src/backend/distributed/deparser/qualify_view_stmt.c:87]: Hmm, in other places like `QualifyRenameTypeStmt` we set `stmt->object = (Node *) names;` not `obj->schemaname`.

I presume it is because relations are represented differently? Is that right?
```C
	RangeVar   *relation;		/* in case it's a table */
	Node	   *object;			/* in case it's some other object */
```

Comment (onderkalaci) [inline src/backend/distributed/deparser/qualify_view_stmt.c:87]: Hmm, in other places like `QualifyRenameTypeStmt` we set `stmt->object = (Node *) names;` not `obj->schemaname`.

I presume it is because relations are represented differently? Is that right?
```C
	RangeVar   *relation;		/* in case it's a table */
	Node	   *object;			/* in case it's some other object */
```

File: src/test/regress/expected/view_propagation.out [hunk 13]
```diff
@@ -359,13 +359,93 @@ ERROR:  cannot change name of view column "id" to "a"
...
+ALTER VIEW alter_view_1 SET (security_barrier);
+ALTER VIEW alter_view_1 SET (check_option=cascaded, security_barrier);
+ALTER VIEW alter_view_1 SET (check_option=cascaded, security_barrier = true);
+ALTER TABLE alter_view_1 SET (check_option=cascaded);
+ALTER TABLE alter_view_1 SET (security_barrier);
+ALTER TABLE alter_view_1 SET (check_option=cascaded, security_barrier);
+ALTER TABLE alter_view_1 SET (check_option=cascaded, security_barrier = true);
+\d+ alter_view_1
+                    View "view_prop_schema.alter_view_1"
+ Column |  Type   | Collation | Nullable | Default | Storage  | Description
```
Comment (onderkalaci) [inline src/test/regress/expected/view_propagation.out:391]: as I also noted in the other PR, `\d` may change outputs per release, so if possible we avoid and query pg_class (such that we wouldn't end up adding new output files if a version changes the description)

Comment (onderkalaci) [inline src/test/regress/expected/view_propagation.out:391]: as I also noted in the other PR, `\d` may change outputs per release, so if possible we avoid and query pg_class (such that we wouldn't end up adding new output files if a version changes the description)

File: src/test/regress/expected/view_propagation.out [hunk 14]
```diff
@@ -359,13 +359,93 @@ ERROR:  cannot change name of view column "id" to "a"
...
+ALTER VIEW alter_view_1 RENAME COLUMN val1 TO val2;
+ALTER VIEW alter_view_1 RENAME val2 TO val1;
+ALTER VIEW alter_view_1 RENAME TO alter_view_2;
+ALTER TABLE alter_view_2 RENAME COLUMN val1 TO val2;
+ALTER TABLE alter_view_2 RENAME val2 TO val1;
+ALTER TABLE alter_view_2 RENAME TO alter_view_1;
+-- Alter owner vith alter view/alter table
+SET client_min_messages TO ERROR;
+CREATE USER alter_view_user;
+SELECT run_command_on_workers($$CREATE USER alter_view_user;$$);
```
Comment (onderkalaci) [inline src/test/regress/expected/view_propagation.out:432]: maybe do `SELECT 1 FROM run_command_on_workers` to avoid enterprise conflicts?

Comment (onderkalaci) [inline src/test/regress/expected/view_propagation.out:432]: maybe do `SELECT 1 FROM run_command_on_workers` to avoid enterprise conflicts?

File: src/test/regress/expected/view_propagation.out [hunk 15]
```diff
@@ -359,13 +359,93 @@ ERROR:  cannot change name of view column "id" to "a"
...
+      run_command_on_workers
+---------------------------------------------------------------------
+ (localhost,57637,t,"CREATE ROLE")
+ (localhost,57638,t,"CREATE ROLE")
+(2 rows)
+
+ALTER VIEW alter_view_1 OWNER TO alter_view_user;
```
Comment (onderkalaci) [inline src/test/regress/expected/view_propagation.out:439]: (a) can you query the view with the non-superuser?  (b) can you create a new view with non=superuser [already added that comment to the other PR, choose which one you want to add the test]?

Those are as important as the rest of the implementation :)

Comment (onderkalaci) [inline src/test/regress/expected/view_propagation.out:439]: (a) can you query the view with the non-superuser?  (b) can you create a new view with non=superuser [already added that comment to the other PR, choose which one you want to add the test]?

Those are as important as the rest of the implementation :)

**Review by onderkalaci — COMMENTED** [review]
Mostly minor comments, now testing the changes (and add comments if I find any issues) .

File: src/backend/distributed/commands/distribute_object_ops.c [hunk 16]
```diff
@@ -868,6 +884,14 @@ static DistributeObjectOps Type_Rename = {
 	.address = RenameTypeStmtObjectAddress,
 	.markDistributed = false,
 };
+static DistributeObjectOps View_Rename = {
```
Comment (onderkalaci) [inline src/backend/distributed/commands/distribute_object_ops.c:887]: There seems to be some issues with `IF EXISTS` in general.

Citus:
```SQL
 ALTER VIEW IF EXISTS v1 rename to t1;
ERROR:  relation "v1" does not exist
```

PG:
```
 ALTER VIEW IF EXISTS v1 rename to t1;
NOTICE:  relation "v1" does not exist, skipping
ALTER VIEW
Time: 1.604 ms
```

Comment (onderkalaci) [reply src/backend/distributed/commands/distribute_object_ops.c:887]: Similarly for set schema, this should not send the command to the workers:
```SQL
alter view if exists sc10."r1 owned view renamed" set schema "r1 owned schama" ;
NOTICE:  relation "r1 owned view renamed" does not exist, skipping
ERROR:  relation "sc10.r1 owned view renamed" does not exist
CONTEXT:  while executing command on localhost:9705
Time: 35.567 ms
```

Can you maybe add set of `IF EXISTS` tests to non-existing tables for all the sub-commands?

File: src/backend/distributed/deparser/deparse_view_stmts.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/deparser/deparse_view_stmts.c:159]: For example,
```SQL

 alter table "r1 owned schama"."r1 owned view 2" ADD COLUMN x int;
ERROR:  unsupported subtype for alter view command
DETAIL:  sub command type: 0

```
Should be as in PG:
```SQL
ERROR:  "r1 owned view 2" is not a table, composite type, or foreign table
```

Is that easy to achieve? I guess even if move deparse to post_process, that should be give the same errors as PG?

File: src/backend/distributed/commands/distribute_object_ops.c (hunk 3 above)
Comment (velioglu) [reply src/backend/distributed/commands/distribute_object_ops.c:1554]: Yes, PG supports altering the name of the column of a view. We've added that check to route alter column name for views and tables properly. Default case has been added to not get compile time warning.

Comment (velioglu) [reply src/backend/distributed/commands/distribute_object_ops.c:1554]: Yes, PG supports altering the name of the column of a view. We've added that check to route alter column name for views and tables properly. Default case has been added to not get compile time warning.

File: src/backend/distributed/deparser/deparse_view_stmts.c (hunk 11 above)
Comment (velioglu) [reply src/backend/distributed/deparser/deparse_view_stmts.c:208]: We add `;` via `AppendAlterViewStmt`. I've also checked other commands.

File: src/backend/distributed/deparser/deparse_view_stmts.c (hunk 2 above)
Comment (velioglu) [reply src/backend/distributed/deparser/deparse_view_stmts.c:159]: Since ALTER VIEW command can handle only the cases above, I've removed throwing error from the default case to let PG to throw the right error locally.

File: src/backend/distributed/deparser/qualify_view_stmt.c (hunk 12 above)
Comment (velioglu) [reply src/backend/distributed/deparser/qualify_view_stmt.c]: Yes, it is right.

File: src/backend/distributed/commands/rename.c [hunk 17]
```diff
@@ -63,14 +64,30 @@ PreprocessRenameStmt(Node *node, const char *renameCommand,
 		return NIL;
 	}
 
-	/* check whether we are dealing with a sequence here */
-	if (get_rel_relkind(objectRelationId) == RELKIND_SEQUENCE)
+	/*
```
Comment (onderkalaci) [inline src/backend/distributed/commands/rename.c:67]: I think we need some more comments here. It would not be clear to the readers that you can do `ALTER TABLE ` on a view. We should mention that (even though we discussed in detail, I was confused while reading the code). At least expand as much as the comment in `PreprocessAlterTableStmt `?

File: src/backend/distributed/commands/distribute_object_ops.c (hunk 16 above)
Comment (onderkalaci) [inline src/backend/distributed/commands/distribute_object_ops.c:887]: maybe add some comments along the lines:
/* PG supports ALTER TABLE on a view as well.
* If user types ALTER VIEW .. RENAME, go through this View_Rename, 
* if users types ALTER TABLE .. RENAME, initially goes through Rename_ANY, which then forwards command to the functions of  View_Rename.
*/


(for other ops as well, because it is really non-trivial to understand this. I wish PG has not allowed this :) )

File: src/backend/distributed/commands/table.c [hunk 18]
```diff
@@ -1939,12 +1968,19 @@ PostprocessAlterTableStmt(AlterTableStmt *alterTableStatement)
...
```
Comment (onderkalaci) [inline src/backend/distributed/commands/table.c:1980]: why do we `copyObject` in other ops but not here? Maybe we should NOT copy at all, because we send workers as `ALTER VIEW` and it is better to run the same on the coordinator?
Not sure, asking your opinion as well

File: src/backend/distributed/commands/view.c [hunk 19]
```diff
@@ -460,3 +425,249 @@ AlterViewOwnerCommand(Oid viewOid)
...
+
+	return viewAddress;
+}
+
+
+/*
+ * IsViewRenameStmt returns whether the passed-in RenameStmt is the following
+ * form:
+ *
+ *   - ALTER VIEW RENAME
```
Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:658]: should we add `ALTER VIEW RENAME COLUMN` as well? because it is not obvious?

File: src/backend/distributed/commands/view.c [hunk 20]
```diff
@@ -460,3 +425,249 @@ AlterViewOwnerCommand(Oid viewOid)
...
+	return NIL;
+}
+
+
+/*
+ * AlterViewSchemaStmtObjectAddress returns the ObjectAddress of the view that is the object
+ * of the alter schema statement.
+ */
+ObjectAddress
+AlterViewSchemaStmtObjectAddress(Node *node, bool missing_ok)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/view.c:620]: can we use `AlterTableSchemaStmtObjectAddress`? (e.g., we need to change 1 assert)

**Review by onderkalaci — APPROVED** [review]
after considering minor style notes, seems ready to go

File: src/backend/distributed/commands/view.c (hunk 20 above)
Comment (velioglu) [reply src/backend/distributed/commands/view.c:620]: I think having an explicit function makes it easier to follow.

File: src/backend/distributed/commands/table.c (hunk 18 above)
Comment (velioglu) [reply src/backend/distributed/commands/table.c:1980]: We copy it from preprocess functions and use the statement directly from postprocess functions. I found it more safe to copy it within preprocess but don't need to copy it from postpocess. So, I'm keeping it as it is.
